### PR TITLE
Fix vendor/k8s.io/client-go/discovery staticcheck

### PR DESCRIPTION
**Pull request by codykaup**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Fixes `vendor/k8s.io/client-go/discovery` staticcheck errors.

**Which issue(s) this PR fixes**:

Part of #92402

**Special notes for your reviewer**:

This change really does two things:

1. Updates our import path of `protobuf` from `github.com/golang/protobuf/proto` (depreciated) to `google.golang.org/protobuf/proto`
2. Bumps `github.com/googleapis/gnostic` from `v0.4.1` to `v0.4.2` which adds a required `ProtoReflect` method ([diff](https://github.com/googleapis/gnostic/compare/v0.4.1...v0.4.2))

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by caesarxuchao** [comment]
/ok-to-test

what was the original staticcheck error?

**Comment by caesarxuchao** [comment]
Also staging/src/k8s.io/apimachinery/go.mod is still using the `github.com/golang/protobuf`. Can you make sure all references to `github.com/golang/protobuf` are updated?

(For the record, https://github.com/golang/protobuf#go-support-for-protocol-buffers suggests using the `google.golang.org/protobuf` module over `github.com/golang/protobuf`)

**Comment by caesarxuchao** [comment]
BTW, did you use the `hack/pin-dependency.sh` and `hack/update-vendor.sh` to update the modules? Those scripts will make sure the update is comprehensive.

**Comment by codykaup** [comment]
> what was the original staticcheck error?

Error from 1:
```
Errors from staticcheck:
vendor/k8s.io/client-go/discovery/discovery_client.go:29:2: package github.com/golang/protobuf/proto is deprecated: Use the "google.golang.org/protobuf/proto" package instead.  (SA1019
```

Error from 2:
```
could not analyze dependency k8s.io/kubernetes/vendor/k8s.io/client-go/discovery of k8s.io/kubernetes/vendor/k8s.io/sample-controller/pkg/generated/clientset/versioned (compile)
vendor/k8s.io/client-go/discovery/discovery_client.go:437:30: cannot use document (variable of type *openapi_v2.Document) as protoreflect.ProtoMessage value in argument to proto.Unmarshal: missing method ProtoReflect (compile)
```
^ Tons of those but bumping the dependency version fixed it right up!

> Also staging/src/k8s.io/apimachinery/go.mod is still using the github.com/golang/protobuf. Can you make sure all references to github.com/golang/protobuf are updated?

I was trying to scope the PR down to the discovery package but I can update that one too! I see there's a few others in various packages, should I clean those up as well?

> BTW, did you use the hack/pin-dependency.sh and hack/update-vendor.sh to update the modules? Those scripts will make sure the update is comprehensive.

I did! I wonder why these tests failed though. I tried running them locally and everything passed. 🤔

**Comment by codykaup** [comment]
I got `bazel` running now and I see the issue. I didn't know that updating the `protobuf` location would also bump the API version from v1 to v2 (TIL). Looks like the biggest question right now is printing the debug message here:

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/apitesting/roundtrip/roundtrip.go#L435

Is there a preference now for how those messages should be printed?

**Comment by codykaup** [comment]
Sorry for all the noise @caesarxuchao! This is ready to go now!

File: go.mod [hunk 1]
```diff
@@ -57,7 +57,7 @@ require (
 	github.com/google/go-cmp v0.4.0
 	github.com/google/gofuzz v1.1.0
 	github.com/google/uuid v1.1.1
-	github.com/googleapis/gnostic v0.4.1
+	github.com/googleapis/gnostic v0.4.2
```
Comment (sttts) [inline go.mod:60]: unrelated change here
